### PR TITLE
feat(validations): implement rand poe and rep poe validation

### LINKED ISSUE #655: Validate RaPoE

### PR DESCRIPTION
**Pull request by tmpolaczyk**
Close #655

### CODE REVIEW

File: validations/src/validations.rs [hunk 1]
```diff
@@ -641,6 +642,23 @@ pub fn validate_candidate(
...
+        max
+    } else {
+        max / total_identities
+    };
+
+    let mut proof: [u8; 32] = [0; 32];
+    proof[0] = (target >> 24) as u8;
+    proof[1] = (target >> 16) as u8;
+    proof[2] = (target >> 8) as u8;
+    proof[3] = target as u8;
```
Comment (aesedepece) [inline validations/src/validations.rs]: Could this be worth a separate function?

**Comment by tmpolaczyk** [comment]
Missing tests for RepPoe

We found a regression in commit mining, when setting `num_witnesses` to 2, the block can be considered valid even though it only has 1 commit.

File: validations/src/validations.rs [hunk 2]
```diff
@@ -874,3 +905,71 @@ impl<'a> UtxoDiff<'a> {
...
+    #[test]
+    fn target_reppoe() {
+        // 100% when the total reputation is 0
+        let max_hash = Hash::with_first_u32(0xFFFF_FFFF);
+        let t00 = calculate_reppoe_threshold(Reputation(0), Reputation(0), 1, 0);
+        let t01 = calculate_reppoe_threshold(Reputation(0), Reputation(0), 100, 0);
+        assert_eq!(t00, max_hash);
+        assert_eq!(t00, t01);
+        let t02 = calculate_reppoe_threshold(Reputation(0), Reputation(0), 1, 1);
+        assert_eq!(t02, Hash::with_first_u32(0xFFFF_FFFF));
```
Comment (lrubiorod) [inline validations/src/validations.rs]:
```suggestion
        assert_eq!(t02, max_hash);
```

File: validations/src/validations.rs [hunk 3]
```diff
@@ -874,3 +905,71 @@ impl<'a> UtxoDiff<'a> {
...
+        // 100% when we have all the reputation
+        let t03 = calculate_reppoe_threshold(Reputation(50), Reputation(50), 1, 1);
+        assert_eq!(t03, Hash::with_first_u32(0xFFFF_FFFF));
```
Comment (lrubiorod) [inline validations/src/validations.rs]:
```suggestion
        assert_eq!(t03, max_hash);
```

File: validations/src/validations.rs [hunk 4]
```diff
@@ -641,6 +655,46 @@ pub fn validate_candidate(
...
+    // This is equivalent to adding 1 reputation to every active identity
+    let total_active_reputation = total_active_reputation.0 + num_active_identities;
+
+    // The number of witnesses for the data request.
+    // If num_witnesses is zero, it will be impossible to commit to this data request
+    // However that is impossible because there is a data request validation that prevents it
+    let num_witnesses = u32::from(num_witnesses);
+
+    let max = u32::max_value();
+    let target = if num_witnesses * my_reputation >= total_active_reputation {
```
Comment (lrubiorod) [inline validations/src/validations.rs]: :speech_balloon:

Comment (aesedepece) [reply validations/src/validations.rs]: :dancer: :dagger:

File: validations/src/validations.rs [hunk 5]
```diff
@@ -874,3 +905,71 @@ impl<'a> UtxoDiff<'a> {
...
+        let t01 = calculate_reppoe_threshold(Reputation(0), Reputation(0), 100, 0);
+        assert_eq!(t00, max_hash);
+        assert_eq!(t00, t01);
+        let t02 = calculate_reppoe_threshold(Reputation(0), Reputation(0), 1, 1);
+        assert_eq!(t02, Hash::with_first_u32(0xFFFF_FFFF));
+        // 100% when we have all the reputation
+        let t03 = calculate_reppoe_threshold(Reputation(50), Reputation(50), 1, 1);
+        assert_eq!(t03, Hash::with_first_u32(0xFFFF_FFFF));
+        // 50% when the total reputation is 1
+        let t03 = calculate_reppoe_threshold(Reputation(0), Reputation(1), 1, 1);
```
Comment (lrubiorod) [inline validations/src/validations.rs]:
```suggestion
        let t04 = calculate_reppoe_threshold(Reputation(0), Reputation(1), 1, 1);
```

File: validations/src/validations.rs [hunk 6]
```diff
@@ -874,3 +905,71 @@ impl<'a> UtxoDiff<'a> {
...
+        assert_eq!(t03, Hash::with_first_u32(0x7FFF_FFFF));
+        // 33% when the total reputation is 1 but there are 2 active identities
+        let t03 = calculate_reppoe_threshold(Reputation(0), Reputation(1), 1, 2);
+        assert_eq!(t03, Hash::with_first_u32(0x5555_5555));
+        // 50% when there are 2 nodes with 50% of the reputation each
+        let t03 = calculate_reppoe_threshold(Reputation(1), Reputation(2), 1, 2);
+        // Since the calculate_reppoe function first divides and later
+        // multiplies, we get a rounding error here
+        assert_eq!(t03, Hash::with_first_u32(0x7FFF_FFFE));
+    }
```
Comment (lrubiorod) [inline validations/src/validations.rs]: Add tests for more witnesses required

File: data_structures/src/chain.rs [hunk 7]
```diff
@@ -373,6 +373,20 @@ impl fmt::Display for Hash {
     }
 }
 
+impl Hash {
+    /// Create a Hash which is all zeros except the first 4 bytes,
+    /// which correspond to the bytes of `x` in big endian
+    pub fn with_first_u32(x: u32) -> Hash {
+        let mut h: [u8; 32] = [0; 32];
```
Comment (lrubiorod) [inline data_structures/src/chain.rs]:
```suggestion
        let mut h: [u8; 32] = [0xFF; 32];
```

**Pull request merged**